packages/zosfiles/src/methods/download/Download.ts

```diff
@@ -275,6 +290,248 @@ export class Download {
         }
     }
 
+    /**
+     * Download data sets that match a DSLEVEL pattern to local files
+     *
+     * @param {AbstractSession}  session      - z/OS MF connection info
+     * @param {string}           patterns     - contains the data set(s) pattern
+     * @param {IDownloadOptions} [options={}] - contains the options to be sent
+     *
+     * @returns {Promise<IZosFilesResponse>} A response indicating the outcome of the API
+     *
+     * @throws {ImperativeError} data set name must be set
+     * @throws {Error} When the {@link ZosmfRestClient} throws an error
+     *
+     * @example
+     * ```typescript
+     *
+     * // Download all "PS" and "PO" datasets that match the pattern "USER.**.DATASET" to "user/data/set/pds/"
+     * await Download.dataSetsMatchingPattern(session, "USER.DATA.SET.PDS");
+     *
+     * // Download all "PS" and "PO" datasets that match the pattern "USER.**.DATASET" to "./path/to/dir/"
+     * await Download.dataSetsMatchingPattern(session, "USER.**.PDS", {directory: "./path/to/dir/"});
+     * ```
+     *
+     * @see https://www.ibm.com/support/knowledgecenter/SSLTBW_2.2.0/com.ibm.zos.v2r2.izua700/IZUHPINFO_API_GetReadDataSet.htm
+     */
+    public static async dataSetsMatchingPattern(session: AbstractSession, patterns: string[],
+        options: IDownloadOptions = {}): Promise<IZosFilesResponse> {
+
+        // Pattern is required to be non-empty
+        ImperativeExpect.toNotBeNullOrUndefined(patterns, ZosFilesMessages.missingPatterns.message);
+        patterns = patterns.filter(Boolean);
+        ImperativeExpect.toNotBeEqual(patterns.length, 0, ZosFilesMessages.missingPatterns.message);
+
+        const result = this.emptyDownloadDsmResult();
+        let zosmfResponses: IZosmfListResponseWithStatus[] = [];
+        const width = 40;
+
+        try {
+            // Get names of all data sets
+            for (const pattern of patterns) {
+                const listOfDataSets = await List.dataSet(session, pattern, { attributes: true });
+                zosmfResponses.push(...listOfDataSets.apiResponse.items);
+            }
+
+            // Check if data sets matching pattern found
+            if (zosmfResponses.length === 0) {
+                return {
+                    success: false,
+                    commandResponse: ZosFilesMessages.noDataSetsMatchingPattern.message,
+                    apiResponse: []
+                };
+            }
+
+            // Exclude archived data sets
```

Comment:
I could be overlooking other logic, but since the 'archived' test comes before the 'exclude' test, might we report that we skipped an archived dataset which the user has already told us to exclude? It seems like excluded datasets would not have to checked for other conditions.

Reply:
I think I now see how datasets would be excluded by the datasetMatching handler before this function does its tests.